cpp/src/IceDiscovery/LookupI.cpp

```diff
@@ -349,27 +334,48 @@ LookupI::findAdapter(const AdapterCB& cb, const std::string& adapterId)
 }
 
 void
-LookupI::foundObject(const Ice::Identity& id, const string& requestId, const Ice::ObjectPrxPtr& proxy)
+LookupI::foundObject(const Ice::Identity& id, const string& requestId, const optional<ObjectPrx>& proxy)
 {
     lock_guard lock(_mutex);
+    if (!proxy)
+    {
+        const Ice::CommunicatorPtr communicator = _lookup->ice_getCommunicator();
+        Warning warn(communicator->getLogger());
+        warn << "ignoring null proxy received by foundObjectById id `"
+            << communicator->identityToString(id) << "' requestId `" << requestId << "'";
+        return;
+    }
+
     map<Ice::Identity, ObjectRequestPtr>::iterator p = _objectRequests.find(id);
-    if(p != _objectRequests.end() && p->second->getRequestId() == requestId) // Ignore responses from old requests
+    // Ignore responses from old requests
+    if (p != _objectRequests.end() && p->second->getRequestId() == requestId)
     {
-        p->second->response(proxy);
+        p->second->response(*proxy);
         _timer->cancel(p->second);
         _objectRequests.erase(p);
     }
 }
 
 void
-LookupI::foundAdapter(const string& adapterId, const string& requestId, const Ice::ObjectPrxPtr& proxy,
-                      bool isReplicaGroup)
+LookupI::foundAdapter(
+    const string& adapterId,
+    const string& requestId,
+    const optional<ObjectPrx>& proxy,
+    bool isReplicaGroup)
 {
     lock_guard lock(_mutex);
+    if (!proxy)
```

Comment:
Ditto. I think it would be cleaner if these APIs took non-optional proxies and we did the throwing in the servant before calling these functions.